src: make OnWorkComplete and OnExecute override-able (#589)
nodejs/node-addon-api: doc/async_worker.md

virtual void Napi::AsyncWorker::OnError(const Napi::Error& e);
```

### OnWorkComplete

> This method is invoked after the work has completed on JavaScript thread.

Reviewer: ```suggestion
This method is invoked after the work has completed on the JavaScript thread.
```

The default implementation of this method checks the status of the work and

> try to dispatch the result to `Napi::AsyncWorker::OnOk` or `Napi::AsyncWorker::Error`

Reviewer: ```suggestion
tries to dispatch the result to `Napi::AsyncWorker::OnOk` or `Napi::AsyncWorker::Error`
```

> if the work has committed an error. If the work was cancelled, neither of

Reviewer: ```suggestion
if the work has committed an error. If the work was cancelled, neither
```

`Napi::AsyncWorker::OnOk` nor `Napi::AsyncWorker::Error` will be invoked.

> After the result dispatched, the default implementation will call into

Reviewer: ```suggestion
After the result is dispatched, the default implementation will call into
```

`Napi::AsyncWorker::Destroy` if `SuppressDestruct()` was not called.

```cpp
virtual void OnWorkComplete(Napi::Env env, napi_status status);
```

### OnExecute

> This method is invoked immediately on the work thread on scheduled.

Reviewer: ```suggestion
This method is invoked immediately on the work thread when scheduled.
```

> The default implementation of this method just call the `Napi::AsyncWorker::Execute`

Reviewer: ```suggestion
The default implementation of this method just calls the `Napi::AsyncWorker::Execute`
```

> The default implementation of this method just calls the `Napi::AsyncWorker::Execute`

Reviewer: ```suggestion
The default implementation of this method just calls the `Napi::AsyncWorker::Execute` method
```

> and handles exceptions if cpp exceptions was enabled.

Reviewer: ```suggestion
and handles exceptions if cpp exceptions were enabled.
```

> 

Reviewer: Maybe add a paragraph here warning sternly against using the `Napi::Env` inside `OnExecute`.

virtual void Napi::AsyncWorker::OnError(const Napi::Error& e);
```

### OnWorkComplete

This method is invoked after the work has completed on JavaScript thread.
The default implementation of this method checks the status of the work and
tries to dispatch the result to `Napi::AsyncWorker::OnOk` or `Napi::AsyncWorker::Error`
if the work has committed an error. If the work was cancelled, neither
`Napi::AsyncWorker::OnOk` nor `Napi::AsyncWorker::Error` will be invoked.
After the result is dispatched, the default implementation will call into
`Napi::AsyncWorker::Destroy` if `SuppressDestruct()` was not called.

```cpp
virtual void OnWorkComplete(Napi::Env env, napi_status status);
```

### OnExecute

This method is invoked immediately on the work thread when scheduled.
The default implementation of this method just calls the `Napi::AsyncWorker::Execute`
and handles exceptions if cpp exceptions were enabled.

> The `OnExecute` method receives an `napi_env` argument. However, do NOT

Reviewer: ```suggestion
The `OnExecute` method receives an `napi_env` argument. However, it must NOT be used
```

> use it within this method, as it does not run on the JavaScript thread and

Reviewer: ```suggestion
within this method, as it does not run on the JavaScript thread and
```
